prusti-viper/src/encoder/procedure_encoder.rs

```diff
@@ -2071,6 +2071,74 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
                             }
                         }
 
+                        "core::slice::<impl [T]>::len" => {
+                            trace!("full_func_proc_name: {}", full_func_proc_name);
+                            trace!("self_ty: {:?}", self_ty);
+                            trace!("def_id: {:?}", def_id);
+                            trace!("substs: {:?}", substs);
+                            trace!("args: {:?}", args);
+                            trace!("destination: {:?}", destination);
+                            trace!("ty: {:?}", ty);
+
+                            assert!(args.len() == 1, "unexpected args to slice::len(): {:?}", args);
+                            let slice_operand = self.mir_encoder.encode_operand_expr(&args[0])
+                                .with_span(span)?;
+
+                            trace!("{:?} := Slice${:?}$len({:?})", destination, ty, slice_operand);
+
+                            // we need to put a label before, it seems..
```

Comment:
Why? Have you tried without? The `len` function doesn't reborrow, so I suspect that the `encode_transfer_args_permissions` and the label are not needed.

Reply:
i get a panic saying "thread 'rustc' panicked at 'Location bb0[6] has not yet been encoded', prusti-viper/src/encoder/procedure_encoder.rs:5612:9".

this is with [this testcase](https://github.com/viperproject/prusti-dev/files/6433678/slice-len.rs.txt) and commenting out the label insertion and `encode_transfer_args_permission` call.

[backtrace.txt](https://github.com/viperproject/prusti-dev/files/6433682/backtrace.txt)

i'm not exactly sure, but it seems to me that as this is the last statement in the basic block some permission transfer stuff is being called anyway, to transfer permissions to the next basic block(s), even though we're not doing anything within the statement itself that would require transfering permissions.

Reply:
adding the label, but leaving out the `encode_transfer_args_permissions` call, i get a fold/unfold-error:
```
error: [Prusti internal error] generating fold-unfold Viper statements failed (FailedToObtain(Pred(LabelledOld("l0", Field(Local(_6: Ref(ref$Slice$i32), Position { line: 3, column: 1, id: 73 }), val_ref: Ref(Slice$i32), Position { line: 3, column: 1, id: 73 }), Position { line: 3, column: 1, id: 73 }), read)))
 --> prusti-tests/tests/verify/pass/slices/slice-len.rs:3:1
  |
3 | / fn slice_len_gt_0(x: &[i32]) {
4 | |     assert!(x.len() >= 0);
5 | | }
  | |_^
```

Reply:
[mir.txt](https://github.com/viperproject/prusti-dev/files/6433699/mir.txt)
[slice-len.rs.vir.txt](https://github.com/viperproject/prusti-dev/files/6433702/slice-len.rs.vir.txt)

Reply:
I see, thanks. It seems required to expire the shared reference that is passed as argument to the MIR function call.